Source/Core/DolphinQt/Config/LogWidget.h

```diff
@@ -47,8 +48,97 @@ class LogWidget final : public QDockWidget, LogListener
 
   QTimer* m_timer;
 
+  // Ring buffer implementation, heavily based on:
+  // https://embeddedartistry.com/blog/2017/4/6/circular-buffers-in-cc#c-=
+  // Modified to be an almost complete drop-in replacement for std::queue
+  // It does not implement some methods compared to std::queue,
+  // and also implements capacity(), clear() and full(), which std::queue does not
+  template <class T>
+  class circular_buffer
+  {
+  public:
+    using value_type = T;
+    using size_type = std::size_t;
+    using reference = T&;
+    using const_reference = const T&;
+
+    explicit circular_buffer(std::size_t count)
+        : m_container(std::make_unique<storage[]>(count)), m_capacity(count)
+    {
+    }
+
+    ~circular_buffer() { clear(); }
+
+    bool empty() const noexcept { return m_head == m_tail && !m_full; }
+    bool full() const noexcept { return m_full; }
+    size_type capacity() const noexcept { return m_capacity; }
+
+    size_type size() const noexcept
+    {
+      size_type size;
+      if (m_full)
+        size = m_capacity;
+      else if (m_head >= m_tail)
+        size = m_head - m_tail;
+      else
+        size = m_capacity + m_head - m_tail;
+      return size;
+    }
+
+    const_reference front() const noexcept
+    {
+      return *reinterpret_cast<const_pointer>(std::launder(&m_container[m_tail]));
+    }
+
+    reference front() noexcept
+    {
+      return const_cast<reference>(const_cast<const circular_buffer*>(this)->front());
+    }
+
+    // If buffer is empty, behaviour is undefined
+    void pop() noexcept
+    {
+      reinterpret_cast<pointer>(std::launder(&m_container[m_tail]))->~value_type();
+
+      m_full = false;
+      m_tail = (m_tail + 1) % m_capacity;
+    }
+
+    template <class... Args>
+    void emplace(Args&&... args)
+    {
+      // If full, delete the element first
+      if (m_full)
+        pop();
+
+      new (&m_container[m_head]) value_type(std::forward<Args>(args)...);
+
+      m_head = (m_head + 1) % m_capacity;
+      m_full = m_head == m_tail;
+    }
+
+    void clear() noexcept
+    {
+      while (!empty())
+        pop();
+
+      m_head = m_tail = 0;
+    }
+
+  private:
+    using storage = std::aligned_storage_t<sizeof(value_type), alignof(value_type)>;
+    using pointer = T*;
+    using const_pointer = const T*;
+
+    std::unique_ptr<storage[]> m_container;
+    std::size_t m_head = 0;
+    std::size_t m_tail = 0;
+    const std::size_t m_capacity;
+    bool m_full = false;
```

Comment:
Maintaining this `m_full` member bothers me. `m_tail` and `m_full` can be replaced by a single `m_size` member. Also, it looks like `pop()` can be implemented in terms of `front()`. The tail calculation should then only be needed in one place (`front()`'s implementation). Or you could even reverse the storage order to only need the calculation on a push. Update `m_size` accordingly. `size()`'s implementation is then simply `return m_size;` instead of an else-if-ladder.

Reply:
I'll look into this.

Reply:
`circular_buffer` is now obsolete, turns out we had a (slightly bugged) `FixedSizeQueue` for the same thing. The only thing it doesn't have compared to `circular_buffer` is construction behaviour - `FixedSizeQueue` holds default-initialized elements, while `circular_buffer` held unallocated buffers like STL.